src/librustdoc/clean/utils.rs

```diff
@@ -176,6 +176,15 @@ crate fn get_real_types(
         // FIXME: remove this whole recurse thing when the recursion bug is fixed
         return res;
     }
+    fn insert(res: &mut FxHashSet<(Type, TypeKind)>, cx: &DocContext<'_>, ty: Type) {
+        if let Some(kind) = ty.def_id().map(|did| cx.tcx.def_kind(did).clean(cx)) {
+            res.insert((ty, kind));
+        } else {
+            // This is a primitive, let's store it as such.
+            res.insert((ty, TypeKind::Primitive));
+        }
+    }
+
```

Comment:
Do you think it makes sense to move this inner function to the start of the enclosing function? It seems confusing to have it wedged in the middle of some code.

Reply:
Oh damn, you're absolutely right. Moving it to the top!